---
title: "1_preprocessing"
output: html_document
date: "2024-12-11"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Import libraries, helper functions, and filepaths

```{r}
library(SoupX)
library(Seurat)
library(DropletUtils)
library(BSgenome.Nfurzeri.NCBI.Nfu20140520.custom)
library(Signac)
library(tidyverse)
library(dplyr)
library(DoubletFinder)
library(scDblFinder)
library(RColorBrewer)
library(scales)
```

```{r}
SaveFigure <- function(plots, name, type = "png", width, height, res){
  if(type == "png") {
    png(paste0(fig_path, name, ".", type),
      width = width, height = height, units = "in", res = 200)
  } else {
    pdf(paste0(fig_path, name, ".", type),
      width = width, height = height)
  }
  print(plots)
  dev.off()
  print(plots)
}

SaveObject <- function(object, name){
  saveRDS(object, paste0(out_path, name, ".rds"))
}

ReadObject <- function(name){
  readRDS(paste0(out_path, name, ".rds"))
}
```

```{r}
#path to spipe output matrix
mat_path <- "../data/all-sample/DGE_filtered/"
#path for saving analysis outputs
out_path <- "../results/"
#path for saving figures
fig_path <- "../results/figures/"

```

## Import matrix

Using the seurat function ReadParseBio, we can load in out cell x gene matrix that we got from spipe. We also can load the metadata and then create a seurat object.

```{r}
mat <- ReadParseBio(mat_path)
cell_meta <- read.csv(paste0(mat_path, "/cell_metadata.csv"), row.names = 1)

#split sample column into temperature, timepoint, and replicate columns
cell_meta <- cell_meta %>%
  separate(col = sample, into = c("temp", "timepoint", "replicate"), sep = "_", remove = F)

so <- CreateSeuratObject(mat, names.field = 0, meta.data = cell_meta)

so
```

```{r}
SaveObject(so, "so_unfiltered")
so <- ReadObject("so_unfiltered")
```

## Identify Doublets with scDblFinder

First, we convert our seurat object to a SingleCellExperiment, which is the format used by scDblFinder

```{r}
#convert seurat object to single cell experiment
sce <- as.SingleCellExperiment(so)
```

Next, we can run scDblFinder, which will automatically use one metadata column to separate samples by, as doublet identification must be done independently one each sample. In this case, we will use the column "sample".

```{r}
#Run scDblFinder
sce <- scDblFinder(sce, samples = "sample")

#View results 
table(sce$scDblFinder.class)

#append to seurat object metadata
doublets <- data.frame(sce@colData)
doublets <- doublets[,(ncol(doublets)-4):ncol(doublets)] #keep columns containing scDblFinder results

so@meta.data <- cbind(so@meta.data, doublets)
```

```{r}
SaveObject(so, "so_post_doubletID")
so <- ReadObject("so_post_doubletID")
```

## QC and Filtering

We will first plot a few different QC metrics to visualize the quality of cells.

Specifically, we are interested in expression of mitochondrial genes, number of genes per cell (nFeature_RNA), and number of counts per cell (nCount_RNA).

We first generate a violin plot showing the distribution of these three metrics. Next, we look at the relationship between number of counts per cell and mitochondrial gene expression and number of genes per cell respectively in scatter plots.

```{r}
#calculate percentage expression of mitochondrial genes, which in the grz genome begin with KEG92
so[["percent.mt"]] <- PercentageFeatureSet(so, pattern = "^KEG92-")

#generate qc plots
p <- VlnPlot(so, pt.size = 0.10, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
SaveFigure(p, "vln_QC", width = 6, height = 6)

p1 <- FeatureScatter(so, feature1 = "nCount_RNA", feature2 = "percent.mt")
p2 <- FeatureScatter(so, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
SaveFigure((p1+p2), "scatter_QC", width = 12, height = 6)

```

Now, we can pick appropriate cutoffs for each of our qc metrics. Note: we have not removed doublets at this step

```{r}
so <- subset(so, subset = nFeature_RNA < 8000 & nCount_RNA < 20000 & percent.mt < 15)
so
```

```{r}
SaveObject(so, "so_filtered")
so <- ReadObject("so_filtered")
```

Next, we run through the standard seurat workflow so we can visualize doublets on a UMAP plot.

```{r}
so <- NormalizeData(so)
so <- FindVariableFeatures(so)
so <- ScaleData(so)
so <- RunPCA(so)
DimPlot(so, reduction = "pca")
so <- RunUMAP(so, dims = 1:30)
```

```{r}
p <- DimPlot(so, reduction = "umap", group.by = "scDblFinder.class", cols = c("grey", "blue"))
SaveFigure(p, "umap_doublets", width = 6, height = 6)

```

Finally, we remove the doublets and save the object.

```{r}
so <- subset(so, scDblFinder.class == "singlet")
so
```

```{r}
SaveObject(so, "so_singlets")
so <- ReadObject("so_singlets")
```

## Seurat Processing

Now, we can re-run the seurat workflow and visualize our data using UMAP plots

```{r}
so <- NormalizeData(so)
so <- FindVariableFeatures(so)
so <- ScaleData(so)
so <- RunPCA(so)
ElbowPlot(so,ndims = 50)
# so <- JackStraw(so, dims = 50, num.replicate = 100)
# so <- ScoreJackStraw(so, dims = 1:50, do.plot = T)
```

```{r}
so <- RunUMAP(so, dims = 1:30)
so <- FindNeighbors(so, dims = 1:30)
so <- FindClusters(so, cluster.name = "louvain")
```

```{r}
p1 <- DimPlot(so, group.by = "louvain", label = T, pt.size = 1) + NoLegend()
SaveFigure(p1, "umap_clusters", width = 10, height = 10)


p2 <- DimPlot(so, group.by = "temp", cols = c("#2171B5", "#D94801"), pt.size = 1)
SaveFigure(p2, "umap_temp", width = 11, height = 10)

warm_colors <- brewer.pal(9, "Oranges")[c(7, 7, 5, 5, 3, 3)]
cool_colors <- brewer.pal(9, "Blues")[c(7, 7, 5, 5, 3, 3)]
color_scheme <- c(cool_colors, warm_colors)
p3 <- DimPlot(so, group.by = "sample", cols = color_scheme, pt.size = 1)
SaveFigure(p3, "umap_sample", width = 11, height = 10)

```

```{r}
#add combined timepoint to metadata
metadata <- so@meta.data
metadata <- metadata %>%
  mutate(timepoint_comb = ifelse(timepoint %in% c("dev1d", "dia6d"), "posthb", timepoint)) %>%
  mutate(timepoint_comb = factor(timepoint_comb, levels = c("posthb", "hb", "predia")))
so@meta.data <- metadata

p <- DimPlot(so, group.by = "sample", split.by = "timepoint_comb", pt.size = 1,
        cols = c("#2171B5", "#2171B5", "#2171B5", "#2171B5", "#2171B5", "#2171B5",
                 "#D94801", "#D94801", "#D94801", "#D94801", "#D94801", "#D94801"))

SaveFigure(p, "umap_timepoint_split", width = 20, height = 10)
```

```{r}
SaveObject(so, "so_processed")
so <- ReadObject("so_processed")
```
